File: ITP_KANDY_7-main/frontend/lib/core/utils/validation_utils.dart
// ------------------------------------------------------------------------------
// File: validation_utils.dart
// Purpose: Deterministic Data Integrity and Business Rule Enforcement.
// Rationale: Centrally orchestrates the logic for verifying user inputs across
//   all system modules. Ensures that only sanitized and valid data is
//   persisted to local state or transmitted to the backend, preventing
//   runtime inconsistencies and financial data corruption.
// ------------------------------------------------------------------------------
class ValidationUtils {
  /*
   * Logic: Domain-Specific Email Validator.
   * Rationale: Limits account registration to @gmail.com domains to simplify 
   *   administrative support and ensure standardized identity providers.
   */
  static String? validateEmail(String? value) {
    if (value == null || value.trim().isEmpty) {
      return null; // Logic: Email is often optional in profile/owner contexts
    }
    final trimmed = value.trim(); // Cleanup: Remove leading/trailing whitespace
    if (!trimmed.toLowerCase().endsWith('@gmail.com')) {
      return 'Email must end with @gmail.com'; // Constraint: Specific domain enforcement
    }
    return null; // Success: Input meets requirements
  }

  /*
   * Logic: Polymorphic Login Identifier Validator.
   * Rationale: Dynamically determines authentication strategy by parsing 
   *   the input signature (Email vs Phone).
   */
  static String? validateIdentifier(String? value) {
    if (value == null || value.trim().isEmpty) {
      return 'Email or Phone is required'; // Constraint: Mandatory login field
    }
    final trimmed = value.trim();
    if (trimmed.contains('@')) {
      return validateEmail(trimmed); // Strategy: Delegate to email logic
    } else {
      return validatePhone(trimmed); // Strategy: Delegate to phone logic
    }
  }

  /*
   * Logic: Regional Phone Validator (Sri Lanka).
   * Rationale: Enforces strict adherence to local (07...) and 
   *   international (+94...) formats to ensure SMS deliverability.
   */
  static String? validatePhone(String? value) {
    if (value == null || value.trim().isEmpty) {
      return 'Phone number is required'; // Constraint: Mandatory for SMS-auth
    }
    final trimmed = value.trim();

    // RegEx: Starts with 0 and followed by 9 decimals.
    final zeroPattern = RegExp(r'^0[0-9]{9}$');
    // RegEx: Starts with +94 and followed by 9 decimals.
    final intlPattern = RegExp(r'^\+94[0-9]{9}$');

    if (trimmed.startsWith('0')) {
      if (!zeroPattern.hasMatch(trimmed)) {
        return 'Phone number must have 9 digits after 0 (e.g. 0771234567)';
      }
    } else if (trimmed.startsWith('+94')) {
      if (!intlPattern.hasMatch(trimmed)) {
        return 'Phone number must have 9 digits after +94 (e.g. +94771234567)';
      }
    } else {
      // Constraint: Reject numbers from unsupported regions or invalid formats.
      return 'Phone number must start with 0 or +94';
    }

    return null;
  }

  /*
   * Logic: Data Security Guard (Entropy Check).
   * Rationale: Enforces a minimum length of 8 characters to mitigate 
   *   brute-force vulnerabilities.
   */
  static String? validatePassword(String? value) {
    if (value == null || value.trim().isEmpty) {
      return 'Password is required'; // Constraint: Mandatory security field
    }
    if (value.trim().length < 8) {
      return 'Password must be at least 8 characters'; // Constraint: Minimum length rule
    }
    return null;
  }

  /*
   * Logic: Global Null/Empty Triage.
   * Rationale: Generic validator for mandatory fields that do not 
   *   require complex pattern matching.
   */
  static String? validateRequired(String? value, String fieldName) {
    if (value == null || value.trim().isEmpty) {
      return '$fieldName is required'; // Dynamic: Custom error based on field role
    }
    return null;
  }

  /*
   * Logic: Identity Validation.
   * Rationale: Enforces a minimum length of 3 characters for names (Customers/Suppliers)
   *   to prevent data ambiguity and accidental single-character entries.
   */
  static String? validateName(String? value, {String fieldName = 'Name'}) {
    if (value == null || value.trim().isEmpty) {
      return '$fieldName is required';
    }
    if (value.trim().length < 3) {
      return '$fieldName must be at least 3 characters';
    }
    return null;
  }

  /*
   * Logic: Financial Data Consistency (Price).
   * Rationale: Prevents zero-value or negative item definitions to ensure 
   *   accurate accounting across Sales and Purchase modules.
   */
  static String? validatePrice(String? value) {
    if (value == null || value.trim().isEmpty) {
      return 'Price is required'; // Constraint: Mandatory for products/orders
    }
    final price = double.tryParse(value); // Attempt: Convert to decimal
    if (price == null || price <= 0) {
      return 'Enter a valid price greater than 0'; // Constraint: Prevent zero/negative pricing
    }
    return null;
  }

  /*
   * Logic: Inventory Data Consistency (Stock).
   * Rationale: Enforces non-negative integer counts for warehouse auditing.
   */
  static String? validateStock(String? value) {
    if (value == null || value.trim().isEmpty) {
      return 'Stock level is required'; // Constraint: Mandatory inventory field
    }
    final stock = int.tryParse(value); // Attempt: Convert to whole number
    if (stock == null || stock < 0) {
      return 'Enter a valid non-negative number'; // Rationale: 0 stock is valid (out of stock)
    }
    return null;
  }
}

File: ITP_KANDY_7-main/backend/src/utils/validationUtils.js
/**
 * Backend Utility: Global Validation Suite.
 * Enforces business-critical data integrity constraints for the ShopBook ecosystem.
 */

/**
 * Logic: Email Integrity.
 * ShopBook currently enforces a Gmail-only policy for merchant accounts for standardized recovery.
 */
const isValidEmail = (email) => {
    if (!email) return true; // Stability: Allow empty if field is optional in specific contexts.
    return email.toLowerCase().endsWith('@gmail.com'); // Constraint: Domain restriction.
};

/**
 * Logic: Contact Channel Verification.
 * Enforces strict formatting for Sri Lankan mobile numbers (Local & International formats).
 */
const isValidPhone = (phone) => {
    if (!phone) return false; // Constraint: Phone is a mandatory identity key.
    const trimmed = phone.trim();
    
    // Pattern A: Standard local 10-digit (e.g., 0771234567, 0741234567)
    // Pattern B: International format with +94 prefix (e.g., +94771234567, +94741234567)
    return /^07[0-9]{8}$/.test(trimmed) || /^\+947[0-9]{8}$/.test(trimmed);
};

/**
 * Logic: Security Baseline.
 * Enforces a minimum entropy standard for merchant passwords.
 */
const isValidPassword = (password) => {
    return password && password.length >= 8; // Constraint: 8-character minimum.
};

/**
 * Logic: Financial Precision.
 * Ensures the provided price is a positive numeric value to prevent accounting errors.
 */
const isValidPrice = (price) => {
    const val = parseFloat(price);
    // Stability: Use isNaN check and ensure it's strictly greater than zero.
    return !isNaN(val) && val > 0;
};

/**
 * Logic: Inventory Accuracy.
 * Ensures stock levels are valid non-negative integers for discrete unit tracking.
 */
const isValidStock = (stock) => {
    // Initial null/undefined check for required inventory fields.
    if (stock === null || stock === undefined || stock === '') return false;
    const val = Number(stock);
    
    // Integers are required for discrete units (e.g., 1 box, 2 items, NOT 1.5).
    return Number.isInteger(val) && val >= 0;
};

// Module Export: Validation logic shared across usecases and infrastructure.
module.exports = {
    isValidEmail,
    isValidPhone,
    isValidPassword,
    isValidPrice,
    isValidStock
};
